Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm

```diff
@@ -138,36 +155,243 @@ - (void)webViewDidClose:(WKWebView *)webView
 
 @end
 
+#if PLATFORM(IOS_FAMILY)
+static void* kvoContext = &kvoContext;
+#endif
+
 @interface _WKWebExtensionActionWebView : WKWebView
+
+@property (nonatomic, readonly) BOOL contentSizeHasStabilized;
+
 @end
 
 @implementation _WKWebExtensionActionWebView {
-    WeakPtr<WebKit::WebExtensionAction> _webExtensionAction;
+    WeakPtr<WebExtensionAction> _webExtensionAction;
+    CGSize _previousContentSize;
 }
 
-- (instancetype)initWithFrame:(CGRect)frame configuration:(WKWebViewConfiguration *)configuration webExtensionAction:(WebKit::WebExtensionAction&)action
+- (instancetype)initWithFrame:(CGRect)frame configuration:(WKWebViewConfiguration *)configuration webExtensionAction:(WebExtensionAction&)action
 {
     if (!(self = [super initWithFrame:frame configuration:configuration]))
         return nil;
 
     _webExtensionAction = action;
 
+#if PLATFORM(IOS_FAMILY)
+    [self.scrollView addObserver:self forKeyPath:@"contentSize" options:NSKeyValueObservingOptionNew context:kvoContext];
+#endif
+
     return self;
 }
 
-- (void)invalidateIntrinsicContentSize
+#if PLATFORM(IOS_FAMILY)
+- (void)dealloc
 {
-    auto intrinsicContentSize = self.intrinsicContentSize;
+    [self.scrollView removeObserver:self forKeyPath:@"contentSize" context:kvoContext];
+}
+#endif
 
+#if PLATFORM(MAC)
+- (void)invalidateIntrinsicContentSize
+{
     [super invalidateIntrinsicContentSize];
 
+    [self _contentSizeDidChange];
+}
+#endif // PLATFORM(MAC)
+
+#if PLATFORM(IOS_FAMILY)
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSString *, id> *)change context:(void*)context
+{
+    if (context != kvoContext) {
+        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
+        return;
+    }
+
+    ASSERT(object == self.scrollView);
+
+    [self _contentSizeDidChange];
+}
+#endif // PLATFORM(IOS_FAMILY)
+
+- (CGSize)_contentSize
+{
+#if PLATFORM(IOS_FAMILY)
+    return self.scrollView.contentSize;
+#else
+    return self.intrinsicContentSize;
+#endif
+}
+
+- (void)_contentSizeDidChange
+{
+    if (!_webExtensionAction)
+        return;
```

Comment:
When would this happen? If the timer fires after the object has been deallocated?

Reply:
Ah - yeah - the `performSelector:afterDelay.`